### PR TITLE
Work around Chisel3's lack of 0-width wires

### PR DESCRIPTION
**Pull request by palmer-dabbelt**
This is super ugly, but it's necessary to get Chisel3 to compile.  Note
that this still fails simulations in Chisel3, so it might be wrong.

### CODE REVIEW

File: src/main/scala/generator.scala [hunk 1]
```diff
@@ -59,10 +59,26 @@ class UncachedTileLinkGenerator(id: Int)
 
   io.finished := (state === s_finished)
 
-  val full_addr = UInt(startAddress) + Cat(
-    req_cnt, UInt(id, log2Ceil(nGens)),
-    (if (genCached) UInt(0, 1) else UInt(0, 0)),
-    UInt(0, wordOffset))
+  print(s"wordOffset: ${wordOffset}\n")
+  print(s"log2Ceil(nGens): ${log2Ceil(nGens)}\n")
+  print(s"wordOffset: ${wordOffset}\n")
```
Comment (aswaterman) [inline src/main/scala/generator.scala]: can you remove these?

**Comment by palmer-dabbelt** [comment]
@aswaterman yes

**Comment by palmer-dabbelt** [comment]
@aswaterman @zhemao I have rocket-chip building upstream with chisel3.  I'm blocked on the rocket-chip/master merge until this gets in  Is this OK?

File: src/main/scala/generator.scala [hunk 2]
```diff
@@ -59,10 +59,23 @@ class UncachedTileLinkGenerator(id: Int)
...
+          UInt(0, wordOffset))
+    }
+  val another_part_of_full_addr =
+    if (log2Ceil(nGens) > 0) {
+      Cat(UInt(id, log2Ceil(nGens)),
+          part_of_full_addr)
+    } else {
+      part_of_full_addr
+    }
+  val full_addr = UInt(startAddress) + another_part_of_full_addr
```
Comment (aswaterman) [inline src/main/scala/generator.scala]: I know it's just a hack around a Chisel3 limitation, but it's really ugly.. can't you do something like

    val full_addr_lsbs =
      if (genCached) Cat(UInt(0, 1), UInt(0, wordOffset))
      else UInt(0, wordOffset)
    val full_addr = UInt(startAddress) + Cat(req_cnt, UInt(id, log2Ceil(nGens)), full_addr_lsbs)

File: src/main/scala/generator.scala [hunk 3]
```diff
@@ -122,10 +135,25 @@ class HellaCacheGenerator(id: Int)
...
+         }
+       } else {
+         if (genUncached) {
+           Cat(UInt(1, width = 1),
+               UInt(0, wordOffset))
+         } else {
+           UInt(0, wordOffset)
+         }
+       }
+  val req_addr = UInt(startAddress) + Cat(req_cnt, part_of_req_addr)
```
Comment (aswaterman) [inline src/main/scala/generator.scala]: Likewise, something like

    val req_addr_lsbs =
      if (genUncached) Cat(UInt(1, 1), UInt(0, wordOffset))
      else UInt(0, wordOffset)
    val req_addr = UInt(startAddress) + Cat(req_cnt, UInt(id, log2Ceil(nGens)), req_addr_lsbs)

**Comment by aswaterman** [comment]
I think you should merge this after taking my line notes into account.

**Comment by palmer-dabbelt** [comment]
log2Ceil(nGens) can be (and in fact, is) 0, so I believe I still need all the cases.  If you want I can collapse them into any number of lines you want.

**Comment by aswaterman** [comment]
oh, I see, it's not just the UInt(0, 0).  My misunderstanding.

**Pull request merged**